### PR TITLE
fix: addToCart takes actual product code into account

### PR DESCRIPTION
**Pull request by ChristophHi**

### CODE REVIEW

File: feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog-event.listener.ts [hunk 1]
```diff
@@ -28,19 +30,33 @@ export class AddedToCartDialogEventListener implements OnDestroy {
...
+          this.closeModal(event);
+        })
+      );
+    }
   }
-
+  /**
+   * Opens modal based on CartUiEventAddToCart.
+   * @param event Signals that a product has been added to the cart.
+   */
```
Comment (Platonn) [inline feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog-event.listener.ts:64]: [MINOR] 
Thank you for adding JSDocs for the old method. What about marking it as @deprecated with a comment recommendation to enable the Feature Toggle `adddedToCartDialogDrivenBySuccessEvent` and use  the method `openModalAfterSuccess` instead? Then let's remove the old method together with the Feature Toggle after 12 months

File: feature-libs/cart/base/core/facade/multi-cart.service.ts [hunk 2]
```diff
@@ -227,7 +227,8 @@ export class MultiCartService implements MultiCartFacade {
     cartId: string,
     productCode: string,
     quantity: number,
-    pickupStore?: string
+    pickupStore?: string,
+    numberOfEntriesBeforeAdd?: number
```
Comment (Platonn) [inline feature-libs/cart/base/core/facade/multi-cart.service.ts]: [MAJOR]
Can you help me understand why we have this new parameter of the method `addEntry()?

IMHO it should not be the responsibility of the consumer of the MultiCartService to tell how many products were there in the cart before the addition. As a consumer, I want to only tell which _new item_ I'm putting into the cart.

Comment (ChristophHi) [reply feature-libs/cart/base/core/facade/multi-cart.service.ts]: Agree to your point. It's better to fetch the number of entries before add in the actual multi-cart.service implementation

File: feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog-event.listener.ts [hunk 3]
```diff
@@ -103,6 +108,19 @@ export class AddedToCartDialogEventListener implements OnDestroy {
     }
   }
 
+  protected calculateEntryWasMerged(
+    successEvent: CartAddEntrySuccessEvent
+  ): boolean {
+    // In case quantityAdded zero, the system could have run into stock issues.
+    // Then we continue and still launch the dialogue in 'merge' mode in order
+    // to not break the existing behaviour.
```
Comment (Platonn) [inline feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog-event.listener.ts]: Nice catch. Nice you handled it!

File: feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog.component.ts [hunk 4]
```diff
@@ -36,12 +36,29 @@ import {
...
+  /**
+   * Number of cart entries before addToCart was triggered.
+   * @deprecated since 2211.24. Enable feature toggle 'adddedToCartDialogDrivenBySuccessEvent'
+   * and use attribute addedEntryWasMerged instead.
+   */
+  numberOfEntriesBeforeAdd?: number;
+  pickupStoreName?: string;
+  /**
+   * Was the product added to the cart merged into an existing cart entry (with increased quantity),
+   * or did the system create a new cart entry?
```
Comment (Platonn) [inline feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog.component.ts]: nitpick: The question mark in the end of the sentence is a bit confusing. What about:
```suggestion
   * Tells whether the product added to the cart was merged into an existing cart entry (with increased quantity),
   * or the system created a new cart entry.
```

File: feature-libs/cart/base/root/events/cart.events.ts [hunk 5]
```diff
@@ -72,6 +72,7 @@ export class CartAddEntrySuccessEvent extends CartEvent {
   entry?: OrderEntry;
   quantityAdded?: number;
   deliveryModeChanged?: boolean;
+  pickupStore?: string;
```
Comment (Platonn) [inline feature-libs/cart/base/root/events/cart.events.ts]: Is this change _really_ related to this PR?

File: projects/core/src/features-config/feature-toggles/config/feature-toggles.ts [hunk 6]
```diff
@@ -43,6 +43,12 @@ export interface FeatureTogglesInterface {
    */
   productConfiguratorAttributeTypesV2?: boolean;
 
+  /**
+   * The addedToCart dialog is driven by 'CartAddEntrySuccessEvent'. Previously it was driven
+   * by 'CartUiEventAddToCart event. Code changes affect 'AddedToCartDialogEventListener'
```
Comment (Platonn) [inline projects/core/src/features-config/feature-toggles/config/feature-toggles.ts]:
```suggestion
   * by 'CartUiEventAddToCart' event. Code changes affect 'AddedToCartDialogEventListener'
```

File: feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog-event.listener.ts [hunk 7]
```diff
@@ -61,6 +82,45 @@ export class AddedToCartDialogEventListener implements OnDestroy {
...
+
+    if (dialog) {
+      dialog.pipe(take(1)).subscribe();
+    }
+  }
+
+  protected calculateEntryWasMerged(
+    successEvent: CartAddEntrySuccessEvent
+  ): boolean {
+    // In case quantityAdded zero, the system could have run into stock issues.
```
Comment (Platonn) [inline feature-libs/cart/base/components/added-to-cart-dialog/added-to-cart-dialog-event.listener.ts]: From unit tests and implementation I understood that in case of stock issues the property `quantityAdded` is `undefined` (not `0`).

So what about updating the comment:
```suggestion
    // In case `quantityAdded` is `undefined`, the system could have run into stock issues.
```

Btw. IMHO we could turn this inline comment in to a fully-fledged JSDoc of this method.

File: feature-libs/cart/base/root/events/cart.events.ts (hunk 5 above)
Comment (ChristophHi) [reply feature-libs/cart/base/root/events/cart.events.ts]: It is related (because we need to display the pickup store name in the confirmation dialogue).
But still this is a good hint, on a second look I saw that this information can also be derived from the cart modification via entry.deliveryPointOfService.name, then the event does not need to be touched!

**Review by Platonn — APPROVED** [review]
